Add native docker images w/ build workflow

joeyparrish opened the pull request:
Rebased, cleaned-up version of PR #292, which adds Docker images that do not rely on Wine, with a GitHub Actions workflow to build and push Docker images to ghcr.io.

Compared with #292, this:
 - is based on 4397ac2 (2024-03-06) instead of d84d456 (2023-07-31)
 - has no changes to newline styles, so the diff is easier to read
 - adds instructions to readme.md on building deps/gcc.Dockerfile
 - fixes case on gcc docker image (must be lowercase apparently)
 - fixes typos
 - streamlines and simplifies the GitHub Actions workflow
 - disables arm64 image builds because they keep hanging in the GitHub Actions environment

Resolves #290
Resolves #304

joeyparrish commented:
I'm debugging and streamlining the workflows on my fork.  I'll convert this to a draft for now until that process is complete.  Sorry for the noise!

joeyparrish commented:
This is ready now.  Please take a look!

Stephane-D commented on makefile.gen:
```diff
@@ -10,18 +10,20 @@ include $(GDK)/common.mk
 SRC_S+= $(wildcard $(SRC)/*.s)
 SRC_S+= $(wildcard $(SRC)/*/*.s)
 SRC_S+= $(wildcard $(SRC)/*/*/*.s)
-SRC_S:= $(filter-out $(SRC)/boot/sega.s,$(SRC_S))
```
Why it was removed ? this line is important to prevent embedding linking sega.s into the lib

joeyparrish replied:
I'm reverting this.  See also my other replies.

joeyparrish replied:
Done.

Stephane-D commented on makefile.gen:
```diff
@@ -94,94 +115,104 @@ Debug: debug
 
-out/sega.o: $(SRC)/boot/sega.s out/rom_head.bin
-	$(CC) -x assembler-with-cpp -Wa,--register-prefix-optional,--bitwise-or $(AFLAGS) -c $(SRC)/boot/sega.s -o $@
+$(OUT)/sega.o: $(OUT)/boot/sega.s $(OUT)/rom_head.bin
```
I don't get from where e get this $(OUT)/boot/sega.s file ?

Stephane-D commented on makefile.gen:
```diff
@@ -94,94 +115,104 @@ Debug: debug
 	$(CC) $(DEFAULT_FLAGS) -c $< -o $@
 
-$(SRC)/boot/sega.s: $(SRC_LIB)/boot/sega.s
+$(OUT)/boot/sega.s: $(SRC_LIB)/boot/sega.s
```
Why did you moved boot/sega.s to _OUT_ folder ? The file make be customized (interrupt callback, bootstrap..) so i think it's important to keep it in _SRC_ folder

joeyparrish replied:
I inherited most of these changes from the original PR.  I just tried to clean it up for review and fix the workflows.

I'm guessing the original author thought that generated files should go to OUT as a rule, and didn't realize that it could be customized.  I would have made the same mistake.

I'll revert this.

joeyparrish replied:
I have run into one problem with these generated sources, though, which I will try to resolve.  If I rebuild SGDK, my existing src/boot/rom_head.c gets overwritten when I rebuild my ROM because the one in SRC_LIB is newer.  I only noticed because I had already customized my rom_head.c file.

joeyparrish replied:
Reverted sega.s moving to OUT.  Fixed overwritten boot files.

Stephane-D commented on Dockerfile:
```diff
@@ -1,36 +1,117 @@
+RUN gcc -fexpensive-optimizations -Os -s src/*.c -o $SGDK_PATH/bin/xgmtool
+
+# Building apj.jar
+WORKDIR $SGDK_PATH/tools/apj/src
+RUN javac sgdk/aplib/*.java
+RUN jar cfe $SGDK_PATH/bin/apj.jar sgdk.aplib.Launcher sgdk/aplib/*.class
+
+# Building lz4w.jar
+WORKDIR $SGDK_PATH/tools/lz4w/src
+RUN javac sgdk/lz4w/*.java
+RUN jar cfe $SGDK_PATH/bin/lz4w.jar sgdk.lz4w.Launcher sgdk/lz4w/*.class
+
+# Building sizebnd.jar
+WORKDIR $SGDK_PATH/tools/sizebnd/src
+RUN javac sgdk/sizebnd/*.java
+RUN jar cfe $SGDK_PATH/bin/sizebnd.jar sgdk.sizebnd.Launcher sgdk/sizebnd/*.class
+
+# Building rescomp.jar
+WORKDIR $SGDK_PATH/tools/rescomp/src
+ENV CLASSPATH="$SGDK_PATH/bin/apj.jar:$SGDK_PATH/bin/lz4w.jar:$SGDK_PATH/tools/rescomp/src"
+RUN cp -r $SGDK_PATH/tools/commons/src/sgdk .
+RUN find . -name "*.java"
+RUN find . -name "*.java" | xargs javac
+RUN echo -e "Main-Class: sgdk.rescomp.Launcher\nClass-Path: apj.jar lz4w.jar" > Manifest.txt
+RUN jar cfm $SGDK_PATH/bin/rescomp.jar Manifest.txt  .
```
Is there any interest in rebuilding the JAR file here instead of just copying them from the SGDK repo ? JAR should be portable.

joeyparrish replied:
Sure, I'll look into it.

joeyparrish replied:
This already seems to rebuild the jar from scratch.  I deleted the jar from my local repo, reran the Docker build, and everything is fine.

With this Dockerfile, I can delete all .exe, .dll, and .jar files from the repo and be perfectly happy in Linux.  Everything built from source, no Wine.  Deleting those from the repo would resolve #297, but I'm not sure if that would have negative consequences for Windows users.  Should I do it here?  In a follow-up PR?  Not at all?

Thanks!

Stephane-D reviewed (changes requested):
See comments in Dockerfile and mekafile.gen

joeyparrish commented:
> Nice, could you add git to the second stage image. It would be handy when using the image in a VS Code devcontainer.

I would like to keep this PR focused.  When it is merged, and we have build pipelines that automatically publish new Docker images, you could always follow up with a PR to add git.  And ultimately, it will be @Stephane-D who decides whether or not to allow that.

joeyparrish commented on makefile.gen:
```diff
@@ -94,94 +115,104 @@ Debug: debug
 
-out/sega.o: $(SRC)/boot/sega.s out/rom_head.bin
-	$(CC) -x assembler-with-cpp -Wa,--register-prefix-optional,--bitwise-or $(AFLAGS) -c $(SRC)/boot/sega.s -o $@
+$(OUT)/sega.o: $(OUT)/boot/sega.s $(OUT)/rom_head.bin
```
This doesn't actually seem to depend on rom_head.bin, so I'm removing that dependency.

joeyparrish replied:
I stand corrected.  sega.s references rom_head.bin.  I didn't realize.  I'll add a comment.

joeyparrish replied:
Done.

joeyparrish commented on makefile.gen:
```diff
@@ -94,94 +115,104 @@ Debug: debug
 
 clean: cleanobj cleanres cleanlst cleandep
-	$(RM) -f out.lst out/cmd_ out/symbol.txt out/rom.nm out/rom.wch out/rom.bin
+	$(RM) -f out.lst $(OUT)/cmd_ $(OUT)/symbol.txt $(OUT)/rom.nm $(OUT)/rom.wch $(OUT)/rom.bin
```
I don't know what out.lst is exactly, but it looks like it belongs in $(OUT), so I'm moving it.

joeyparrish replied:
Done.

joeyparrish commented on Dockerfile:
```diff
@@ -1,36 +1,117 @@
-ENV GDK=/sgdk
-ENV SGDK_DOCKER=y
+WORKDIR /sgdk
+COPY . .
```
I'm reorganizing the steps so more of them can be cached.  When I change any one file in SGDK and rebuild, everything past this point must be rerun because the source folder changed and the cache of this step and those after it become stale.

joeyparrish replied:
Done.

joeyparrish commented:
@Stephane-D, I believe I have addressed all the feedback so far.  My only outstanding question for you is this:

This Dockerfile can regenerate all binaries.  I can delete all .exe, .dll, .jar, and .a files from the repo and everything still works from source natively on Linux.  Deleting those from the repo would resolve https://github.com/Stephane-D/SGDK/issues/297, but I'm not sure if that would have negative consequences for Windows users. Should I do it here? In a follow-up PR? Not at all?

Thanks!

Stephane-D commented:
> @Stephane-D, I believe I have addressed all the feedback so far. My only outstanding question for you is this:
> 
> This Dockerfile can regenerate all binaries. I can delete all .exe, .dll, .jar, and .a files from the repo and everything still works from source natively on Linux. Deleting those from the repo would resolve #297, but I'm not sure if that would have negative consequences for Windows users. Should I do it here? In a follow-up PR? Not at all?
> 
> Thanks!

Thanks for taking the time to fix all that !
Well about the JAR file rebuild, it's just that i'm afraid that this increase significantly the Docker call. As I understand it, it will rebuild all tools at each time will compile through the Docker call right ?

joeyparrish commented:
> Well about the JAR file rebuild, it's just that i'm afraid that this increase significantly the Docker call. As I understand it, it will rebuild all tools at each time will compile through the Docker call right ?

Those steps (building tools, jars, etc) will only run when someone runs "docker build ..." to create the image from SGDK source.  The GitHub Actions workflow will build images every time you push to the `master branch`.  These are all the steps with "RUN".

When a user wants to compile their ROM with SGDK using Docker, the tools will not be rebuilt.  It will start at "ENTRYPOINT".

Does this answer your concern?

joeyparrish commented:
@Stephane-D, let me know if you have other concerns or if you would like me to make other changes.  I believe the PR is in a good state.  Everything is rebuilt from source at Docker build time in the automated workflow.  Each change to the `master` branch will trigger a new Docker image with native Linux binaries to be published via GitHub.  And the Docker image is now based on Alpine Linux instead of Ubuntu, which should make it much smaller and faster to download.

iratahack commented:
@joeyparrish , any thoughts on using [crosstool-ng](https://crosstool-ng.github.io/) and bulding the latest gcc 13.2.0?

joeyparrish commented:
> @joeyparrish , any thoughts on using [crosstool-ng](https://crosstool-ng.github.io/) and bulding the latest gcc 13.2.0?
> 
> Also, Sjasm sources are included under tools as a submodule. There is no need to clone independently.

Looks like it would be useful.  I could explore redoing the GCC Dockerfile with crosstool-ng after this PR is merged.  Or you could send a PR yourself if you already have experience with crosstool-ng.  (I have none.)

iratahack commented:
> > @joeyparrish , any thoughts on using [crosstool-ng](https://crosstool-ng.github.io/) and bulding the latest gcc 13.2.0?
> > Also, Sjasm sources are included under tools as a submodule. There is no need to clone independently.
> 
> Looks like it would be useful. I could explore redoing the GCC Dockerfile with crosstool-ng after this PR is merged. Or you could send a PR yourself if you already have experience with crosstool-ng. (I have none.)

I have the Dockerfile specific to building the tools using crosstool-ng but I may need some assistance integrating it. Maybe a collaboration? Take a look at the Dockerfile here https://github.com/iratahack/SGDK/commit/ba9fc473937e2d79ab349940dc1d5272325910f1

joeyparrish commented:
@iratahack, I was hoping to get this PR merged as a baseline before I moved on to improvements, so I would at least be able to refer to the official Docker image from ghcr.io for my own project.  But since this hasn't been re-reviewed, I'll take a look at your work and try to integrate it if I can.  It may be days to weeks before I have the time, though.  Thank you for your offer of collaboration!
